src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/MapRequestPayloadValueResolver.php

```diff
@@ -0,0 +1,130 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel\Controller\ArgumentResolver;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Attribute\MapQueryString;
+use Symfony\Component\HttpKernel\Attribute\MapRequestBody;
+use Symfony\Component\HttpKernel\Controller\ValueResolverInterface;
+use Symfony\Component\HttpKernel\ControllerMetadata\ArgumentMetadata;
+use Symfony\Component\HttpKernel\Exception\HttpException;
+use Symfony\Component\Serializer\Exception\NotEncodableValueException;
+use Symfony\Component\Serializer\Normalizer\AbstractObjectNormalizer;
+use Symfony\Component\Serializer\Normalizer\DenormalizerInterface;
+use Symfony\Component\Serializer\SerializerInterface;
+use Symfony\Component\Validator\Exception\ValidationFailedException;
+use Symfony\Component\Validator\Validator\ValidatorInterface;
+
+/**
+ * @author Konstantin Myakshin <molodchick@gmail.com>
+ */
+final class MapRequestPayloadValueResolver implements ValueResolverInterface
+{
+    private const DEFAULT_FORMAT = 'json';
+
+    private const CONTEXT_DENORMALIZE = [
+        AbstractObjectNormalizer::DISABLE_TYPE_ENFORCEMENT => true,
+        DenormalizerInterface::COLLECT_DENORMALIZATION_ERRORS => true,
+    ];
+
+    private const CONTEXT_DESERIALIZE = [
+        DenormalizerInterface::COLLECT_DENORMALIZATION_ERRORS => true,
+    ];
+
+    public function __construct(
+        private readonly ?SerializerInterface $serializer,
+        private readonly ?DenormalizerInterface $denormalizer,
+        private readonly ?ValidatorInterface $validator,
+    ) {
+    }
+
+    public function resolve(Request $request, ArgumentMetadata $argument): iterable
+    {
+        $payloadMappers = [
+            MapQueryString::class => $this->mapQueryString(...),
+            MapRequestBody::class => $this->mapRequestBody(...),
+        ];
+
+        $payloads = [];
+        foreach ($payloadMappers as $mappingAttribute => $payloadMapper) {
+            $attributes = $argument->getAttributesOfType($mappingAttribute);
+
+            if (!$attributes) {
+                continue;
+            }
+
+            $type = $argument->getType();
+            if (!$type) {
+                throw new \LogicException(sprintf('Could not resolve the "$%s" controller argument: argument should be typed.', $argument->getName()));
+            }
+
+            $payload = $payloadMapper($request, $type, $attributes[0]);
+
+            if ($this->validator) {
+                $violations = $this->validator->validate($payload);
+
+                if (\count($violations)) {
+                    throw new ValidationFailedException($payload, $violations);
+                }
+            }
```

Comment:
What about merging PartialDenormalizationException and other regular exception? I really think it'll be better for the end user.

You can find some code in [my PR](https://github.com/symfony/symfony/pull/42502) about this. And if you want, I can do it

Reply:
Do you mean to get rid of the separate `PartialDenormalizationExceptionListener` and inline try-catch inside value resolver?

BTW `PartialDenormalizationException` works little bit buggy IMHO. We have a lot of `TypeErrors` instead of `PartialDenormalizationException`. Or sometimes we have completely invalid `propertyPath`. You can play with functional test in the current PR to prove that. Any chance to improve it?

Reply:
See https://github.com/symfony/symfony/pull/49138#discussion_r1090546327 for more info

Reply:
I replied for the comment, but anyway, even if a bug exist, it's better IMHO to merge all validation errors

Reply:
Not sure that I got what you mean by "merge all validation errors". Feel free to push you changes to my branch directly/open PR to my fork/provide patch with suggested changes

Reply:
I'll do a PR once merged 👍🏼

Reply:
https://github.com/symfony/symfony/pull/50019 and https://github.com/symfony/symfony/pull/50024